riscv/src/small_field/code_gen.rs

```diff
@@ -0,0 +1,1793 @@
+        runtime.submachines_declare(),
+        preamble,
+        initial_memory
+            .into_iter()
+            .format_with(",\n", |line, f| f(&format_args!("\t\t{line}"))),
+        program
+            .into_iter()
+            .format_with("\n", |line, f| f(&format_args!("\t\t{line}"))),
+    )
+}
+
+fn preamble(field: KnownField, runtime: &Runtime, with_bootloader: bool) -> String {
+    let bootloader_preamble_if_included = if with_bootloader {
+        bootloader_preamble(field)
+    } else {
+        "".to_string()
+    };
+
+    for machine in [
+        "binary",
+        "shift",
+        "bit2",
+        "bit6",
+        "bit7",
+        "byte",
+        "bit12",
+        "byte2",
+        "byte_binary",
+        "byte_shift",
+        "byte_compare",
+    ] {
+        assert!(
+            runtime.has_submachine(machine),
+            "RISC-V machine requires the `{machine}` submachine"
+        );
+    }
+
+    let mul_instruction = mul_instruction();
+
+    r#"
+    reg pc[@pc];
+    reg XL[<=];
+    reg XH[<=];
+    reg YL[<=];
+    reg YH[<=];
+    reg ZL[<=];
+    reg ZH[<=];
+    reg WL[<=];
+    reg WH[<=];
+"#
+        .to_string()
+        // runtime extra registers
+        + &runtime
+            .submachines_extra_registers()
+            .into_iter()
+            .map(|s| format!("\t\t{s}\n"))
+            .join("")
+        + &bootloader_preamble_if_included
+        + &memory(with_bootloader)
+        + r#"
+    // =============== Register memory =======================
+"# + "std::machines::small_field::memory::Memory regs(bit12, byte2);"
+        + r#"
+    // Get the value in register Y.
+    instr get_reg YL -> XH, XL link ~> (XH, XL) = regs.mload(0, YL, STEP);
+
+    // Set the value in register X to the value in register Y.
+    instr set_reg XL, YH, YL -> link ~> regs.mstore(0, XL, STEP, YH, YL);
+
+    // We still need these registers prover inputs.
+    reg query_arg_1_l;
+    reg query_arg_1_h;
+    reg query_arg_2_l;
+    reg query_arg_2_h;
+
+    // Witness columns used for intermediate values inside instructions.
+    col witness tmp1_h;
+    col witness tmp1_l;
+    col witness tmp2_h;
+    col witness tmp2_l;
+    col witness tmp3_h;
+    col witness tmp3_l;
+    col witness tmp4_h;
+    col witness tmp4_l;
+    col witness tmp5_h;
+    col witness tmp5_l;
+    col witness tmp6_h;
+    col witness tmp6_l;
+
+    link => byte2.check(tmp1_h);
+    link => byte2.check(tmp1_l);
+    link => byte2.check(tmp2_h);
+    link => byte2.check(tmp2_l);
+    link => byte2.check(tmp3_h);
+    link => byte2.check(tmp3_l);
+    link => byte2.check(tmp4_h);
+    link => byte2.check(tmp4_l);
+    link => byte2.check(tmp5_h);
+    link => byte2.check(tmp5_l);
+    link => byte2.check(tmp6_h);
+    link => byte2.check(tmp6_l);
+
+    // We need to add these inline instead of using std::utils::is_zero
+    // because when XX is not constrained, witgen will try to set XX,
```

Comment:
Why was this not a problem in the other machine?

Reply:
You could also just add a hint.

Reply:
> You could also just add a hint.

I think we tried this while prototyping the machine and for some reason it didn't work. Can you give me an example in the asm machine?